repair: Switch to use NODE_OPS_CMD for replace operation

asias opened the pull request:
In commit c82250e0cfacb62617155ac98583ac90aed40133 (gossip: Allow deferring
advertise of local node to be up), the replacing node is changed to postpone
the responding of gossip echo message to avoid other nodes sending read
requests to the replacing node. It works as following:

1) replacing node does not respond echo message to avoid other nodes to
   mark replacing node as alive

2) replacing node advertises hibernate state so other nodes knows
   replacing node is replacing

3) replacing node responds echo message so other nodes can mark
   replacing node as alive

This is problematic because after step 2, the existing nodes in the
cluster will start to send writes to the replacing node, but at this
time it is possible that existing nodes haven't marked the replacing
node as alive, thus failing the write request unnecessarily.

For instance, we saw the following errors in issue #8013 (Cassandra
stress fails to achieve consistency when only one of the nodes is down)

```
scylla:
[shard 1] consistency - Live nodes 2 do not satisfy ConsistencyLevel (2
required, 1 pending, live_endpoints={127.0.0.2, 127.0.0.1},
pending_endpoints={127.0.0.3}) [shard 0] gossip - Fail to send
EchoMessage to 127.0.0.3: std::runtime_error (Not ready to respond
gossip echo message)

c-s:
java.io.IOException: Operation x10 on key(s) [4c4f4d37324c35304c30]:
Error executing: (UnavailableException): Not enough replicas available
for query at consistency QUORUM (2 required but only 1 alive
```

To solve this problem, we can do the replacing operation in multiple stages.

One solution is to introduce a new gossip status state as proposed
here: gossip: Introduce STATUS_PREPARE_REPLACE #7416

1) replacing node does not respond echo message

2) replacing node advertises prepare_replace state (Remove replacing
   node from natural endpoint, but do not put in pending list yet)

3) replacing node responds echo message

4) replacing node advertises hibernate state (Put replacing node in
   pending list)

Since we now have the node ops verb introduced in
829b4c14380020fa4058335c4c43cefec135b3de (repair: Make removenode safe
by default), we can do the multiple stage without introducing a new
gossip status state.

This patch uses the NODE_OPS_CMD infrastructure to implement replace
operation.

Improvements:

1) It solves the race between marking replacing node alive and sending
   writes to replacing node

2) The cluster reverts to a state before the replace operation
   automatically in case of error. As a result, it solves when the
   replacing node fails in the middle of the operation, the repacing
   node will be in HIBERNATE status forever issue.

3) The gossip status of the node to be replaced is not changed until the
   replace operation is successful. HIBERNATE gossip status is not used
   anymore.

4) Users can now pass a list of dead nodes to ignore explicitly.

Fixes #8013

avikivity commented on idl/partition_checksum.idl.hh:
```diff
@@ -110,13 +110,20 @@ enum class node_ops_cmd : uint32_t {
+     replace_heartbeat,
+     replace_abort,
+     replace_done,
 };
```
Do we need a cluster feature to prevent the new commands being sent to nodes that don't recognize them?

Probably, it's not critical since this is a rare operation, but we need to make sure new nodes respond correctly to responses from old nodes.

asias replied:
Yes, old olds are handled below.

avikivity commented on idl/partition_checksum.idl.hh:
```diff
@@ -110,13 +110,20 @@ enum class node_ops_cmd : uint32_t {
     utils::UUID ops_uuid;
     std::list<gms::inet_address> ignore_nodes;
     std::list<gms::inet_address> leaving_nodes;
+    std::unordered_map<gms::inet_address, gms::inet_address> replace_nodes;
```
I am guessing that the key is the node being replaced and the value is the node replacing, but I should be reading a comment, not guessing.

avikivity commented on repair/repair.hh:
```diff
@@ -498,6 +504,8 @@ struct node_ops_cmd_request {
     std::list<gms::inet_address> ignore_nodes;
     std::list<gms::inet_address> leaving_nodes;
+    // Map existing nodes to replacing nodes
+    std::unordered_map<gms::inet_address, gms::inet_address> replace_nodes;
```
Ah, here is the comment. But it should also be present in the IDL.

asias replied:
OK.  I will add to IDL too.

avikivity commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,138 @@ future<> storage_service::decommission() {
 
+// Runs inside seastar::async context
+void storage_service::run_replace_ops() {
+    if (!db().local().get_replace_address()) {
```
(replace_address doesn't belong in `database`, that shouldn't be aware of clustering)

asias replied:
This is historical and derived from C*.

avikivity replied:
Yes, I put the comment in parenthesis to indicate it's not on the patch.

avikivity commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,138 @@ future<> storage_service::decommission() {
+            return _messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
+                slogger.debug("replace[{}]: Got node_ops_cmd::replace_prepare response from node={}", uuid, node);
+            }).handle_exception_type([&nodes_unknown_verb, node, uuid] (seastar::rpc::unknown_verb_error&) {
+                slogger.warn("replace[{}]: Node {} does not support node_ops_cmd verb", uuid, node);
```
Ah, I see you handle old nodes.

asias replied:
Yes.

avikivity commented:
Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?

asias commented:
> Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?

This patch is more about the control plane. The data plane that uses repair to sync data does not change. I have tested with dtest. This patch does not solve any latency issues during replace.

asias commented:
Changes in v2:
- Add comments in IDL for the replace_nodes field in node_ops_cmd_request.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,138 @@ future<> storage_service::decommission() {
+        req.cmd = node_ops_cmd::replace_prepare_mark_alive;
+        parallel_for_each(sync_nodes, [this, &req, uuid] (const gms::inet_address& node) {
+            return _messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
+                slogger.debug("replace[{}]: Got repare_mark_alive response from node={}", uuid, node);
```
repare -> prepare

asias replied:
fixed.

asias commented on service/storage_service.cc:
```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                        auto existing_node = x.first;
+                        auto replacing_node = x.second;
+                        slogger.info("replace[{}]: Added replacing_node={} to replace existing_node={}, coordinator={}", req.ops_uuid, replacing_node, existing_node, coordinator);
+                        tmptr->add_replacing_endpoint(existing_node, replacing_node);
```
This is the key point in the solution. After adding replacing node, the replacing node will be excluded from the normal read and write, i.e., other nodes will not send either read or write requests to replacing node.

tgrabiec replied:
What if that node restarts after we executed this step? Will it revert to the previous state?

asias replied:
The participant node do not remember the state, so no. But if the node restarts, the replacing operation will fail anyway. The cluster will revert. So the node does not need to revert to the state with add_replacing_endpoint executed.

tgrabiec replied:
What guarantees that the operation will fail?

Before it fails, is it possible that the node which restarted will use the wrong (old) topology, and will try to read from the new node as if it was the old node being replaced?

asias replied:
The rpc connections to the replacing node will be killed when the replacing node detects a node is restarted, so the streaming will fail and in turn the replace operation will fail.

If the streaming process is already finished before the node restarts, it is not a big problem because the replacing node already have all the data. 

In addition, since the restarted node does not know the ops_uuid, the restarted node will fail any cmd sent by the replacing node, in turn fails the node operation.

tgrabiec replied:
What if the node restarts before we start streaming?

What about the second concern about consistency?

asias replied:
```
In addition, since the restarted node does not know the ops_uuid, the restarted node will fail any cmd sent by the replacing node, in turn fails the node operation.
```

See v2.

tgrabiec replied:
The replace operation will eventually fail, but we still will run into consistency problems until it does, because the replacing node will respond to requests as if it was the old node, sent by the node which restarted and is using the old topology.

The problem here is that the state transitions are not monotonic, restarts revert them, but the solution depends on them being monotonic. If status was propagated via gossip, they would be monotonic.

asias replied:
The problem with gossip propagation can be even worse, e.g., the other node can completely miss the gossip propagation.

So the problem here is

- replacing node uses the same ip address of the node to be replaced
- in the middle of the replace operation one of the node restarted and the restarted node will mark the replacing node as up

Let's break it down:

// Step 1: Decide who needs to sync data for replace operation
// Step 2: Prepare to sync data
// Step 3: Start heartbeat updater
// Step 4: Allow nodes to mark the replacing node as alive
_gossiper.advertise_myself().get();
// Step 5: Wait for nodes to finish marking the replacing node as live
// Step 6: Update pending ranges on nodes
// Step 7: Sync data for replace
// Step 8: Finish

If the restart happens before Step 4, it is not a problem. The replacing node will not respond to mark alive at all. 

If the restart happens after Step 4, we already have the heartbeat updater, The restarted node will fail the heart beat update. We can fail the replace operation.

tgrabiec replied:
I don't think you addressed my concern about potential consistency problems. Even if replace operation eventually fails, some requests may use the wrong topology in the mean time.

asias replied:
See v3. The restarted node will not mark the replacing node as alive.

asias commented on service/storage_service.cc:
```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                if (req.replace_nodes.size() > 1) {
+                    auto msg = format("replace[{}]: Could not replace more than one node at a time: replace_nodes={}", req.ops_uuid, req.replace_nodes);
+                    slogger.warn("{}", msg);
+                    throw std::runtime_error(msg);
```
In general yes, but in this case, it is helpful to log on the receiver side too.

tgrabiec replied:
I don't see it. Not that you have to convince me, just curios.

asias replied:
Replacing multiple nodes in one shot is  not supposed to happen. If this happen, something goes wrong. Logging this rare event on all participant nodes helps understanding the state of the cluster more easily.

tgrabiec replied:
Isn't it enough that it's logged on the initiating node? That's where you start looking when the op fails.

asias replied:
Sometimes you do not have all the logs from the cluster. It is also easier when you see it explicitly from the node rejecting the requests when something went wrong. At times, it is not easy to figure out which node the exception comes from.

asias commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,138 @@ future<> storage_service::decommission() {
+                slogger.debug("replace[{}]: Got abort response from node={}", uuid, node);
+            });
+        }).get();
+        slogger.info("replace[{}]: Aborted replace operation, replace_nodes={}, sync_nodes={}, ignore_nodes={}", uuid, replace_nodes, sync_nodes, ignore_nodes);
```
Since we can not recover from this error, yes, error is a better logger level. The abort handler is non blocking, so it is supposed to be fast to finish.

tgrabiec replied:
send_node_ops_cmd() can block, e.g. until it times out

asias replied:
Yes, every rpc verb can block. But so what? Do we need to log every call before and after each verb.

tgrabiec replied:
It may be useful to know when the abort sequence was started, before we issued the first abort request.

asias replied:
All right, let's add one log at the beginning. It is not a big deal.

asias commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,138 @@ future<> storage_service::decommission() {
+                        return make_ready_future<>();
+                    });
+                }).handle_exception([uuid] (std::exception_ptr ep) {
+                    slogger.warn("replace[{}]: Failed to send heartbeat", uuid);
```
done

asias commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,138 @@ future<> storage_service::decommission() {
+                return make_ready_future<>();
+            }
+            return _messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
+                slogger.debug("replace[{}]: Got abort response from node={}", uuid, node);
```
Why? The node will shutdown itself.

tgrabiec replied:
Is there no problem with leaving some of the nodes not aborted?

asias replied:
No. I think you probably missed the auto revert part of this patch. If no heartbeat is sent by the replacing node, other nodes will revert automatically after some time.

tgrabiec replied:
Yes, I missed that.

However, even if the node reverts automatically, there will still be some delay until that happens (30s?). Admin is not supposed to run any other topology change during that period, right? Otherwise, we run into the concurrent topology changes problem. 

Also, what happens if you restart the replace after such a failed abort, won't the watchdog cancellation interfere with the new operation by modifying token metadata?

asias replied:
Yes, we can not run multiple topology change at the same time. I am thinking about make it more strict by failing the ops attempt if there is another ops id. This way the admin will know for sure that there is another ops going on.

tgrabiec replied:
Sounds good

asias commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,138 @@ future<> storage_service::decommission() {
+        slogger.debug("replace[{}]: Check node={}, status={}", uuid, node, _gossiper.get_gossip_status(node));
+        if (node != get_broadcast_address() &&
+                node != replace_address &&
+                (_gossiper.is_normal(node) || _gossiper.is_shutdown(node)) &&
```
I did not use get_live_token_owners  for various reasons. 1) we do not care about / want to depend on gossip liveness here because we will send verb to those nodes anyway 2) Check normal and shutdown status directly is better than checking is_dead_state. If a node is in the ring, it should be either normal or shutdown.

tgrabiec replied:
Regardless of how the check should look like, we should extract it so that it's clear what we intend it to mean, e.g `is_ring_member()`. 

If we have that, then it's better if all the places which share the meaning reuse this function, because then there is no duplication of logic and there is a higher chance that the whole system remains correct now and in the future.

asias replied:
ok. I will add a helper.

asias commented on service/storage_service.cc:
```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                }).get();
+            } else if (req.cmd == node_ops_cmd::replace_heartbeat) {
+                slogger.debug("replace[{}]: Updated heartbeat from coordinator={}", req.ops_uuid, coordinator);
+                ss.node_ops_update_heartbeat(ops_uuid);
```
The idea is to avoid introducing verbs more than necessary.

tgrabiec replied:
I meant new node_ops_cmd (not RPC verb). We don't need separate cmds for every op, e.g. `removenode_done` and `replace_done`, since both just call "done" on a given op. So you can have a single command called  "done".

asias replied:
For now, the handlers are same for removenode and replace, but in the future we might want to handle them differently, e.g., count how many ops were performed and when was finished.  It is better to future proof. The benefit of saving node_ops_cmd enum is tiny.

tgrabiec replied:
The information about the kind of the operation is already known on the receiver, it is associated with a given uuid when the op is constructed. So you can still count per-op things even if we have a single command to terminate a given op.

It's not a big deal, it only seems like a duplication of effort to implement each of these separately.

asias replied:
When a node restarts, it lost the uuid to ops mapping.

asias commented on service/storage_service.cc:
```diff
@@ -682,6 +682,8 @@ void storage_service::bootstrap() {
         set_mode(mode::JOINING, sprint("Wait until local node knows tokens of peer nodes"), true);
         _gossiper.wait_for_range_setup().get();
+
+      if (!is_repair_based_node_ops_enabled()) {
```
no

tgrabiec replied:
So we cannot mark the issue as fixed by this PR.

asias replied:
Well, I doubt we will fix the path with repair based node ops disabled.

tgrabiec replied:
We could change storage service listeners to wait for mark_alive pings before they change the token metadata.

asias replied:
The listeners runs inside gossip loop. We should not block the loop waiting for the ping to successful.

Alternatively, we can use all the code path here but use streaming for Step 7 to sync data instead using repair to sync.

tgrabiec replied:
The storage service listeners could move to background (serializing) and don't block the gossiper.

asias replied:
> > > Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?
> > 
> > 
> > This patch is more about the control plane. The data plane that uses repair to sync data does not change. I have tested with dtest. This patch does not solve any latency issues during replace.
> 
> Does it mean that the decision whether to use repair or not does not change (and it will still not be used by default?)



> > > Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?
> > 
> > 
> > This patch is more about the control plane. The data plane that uses repair to sync data does not change. I have tested with dtest. This patch does not solve any latency issues during replace.
> 
> Does it mean that the decision whether to use repair or not does not change (and it will still not be used by default?)



> > > Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?
> > 
> > 
> > This patch is more about the control plane. The data plane that uses repair to sync data does not change. I have tested with dtest. This patch does not solve any latency issues during replace.
> 
> Does it mean that the decision whether to use repair or not does not change (and it will still not be used by default?)

Yes, exactly.

asias commented on service/storage_service.cc:
```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                            req.ops_uuid, req.replace_nodes, replacing_nodes_live);
+                    if (replacing_nodes_live.size() == nr_replacing_nodes) {
+                        break;
+                    }
```
Here we wait for more than one nodes. We also want to know which nodes are alive for logging. Maybe a follow up patch.

tgrabiec replied:
You can call wait_alive() inside parallel_for_each().

You can log between wait_alive() calls.

asias replied:
If I need to collect nodes from gossip then call wait_alive for each node, why not call the existing is_alive in the parallel_for_each directly.  Also, we  want to check the total time spent waiting not the individual node.

tgrabiec replied:
The point is to make the code cleaner, easier to understand, by abstracting details into higher-level functions (like wait_alive()). It's more important that reducing the number of function calls made.

I don't see why my suggestion prohibits you from measuring the total time spent waiting.

asias replied:
I did not say it prohibits. It makes it more complicated to count each node separately. I will make it a helper which works on multiples nodes.

avikivity commented:
> > Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?
> 
> This patch is more about the control plane. The data plane that uses repair to sync data does not change. I have tested with dtest. This patch does not solve any latency issues during replace.

Yes, but won't any unresolved latency problems in the data plane be exposed as a result of the patch?

asias commented:
> > > Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?
> > 
> > 
> > This patch is more about the control plane. The data plane that uses repair to sync data does not change. I have tested with dtest. This patch does not solve any latency issues during replace.
> 
> Yes, but won't any unresolved latency problems in the data plane be exposed as a result of the patch?

This patch does not enable repair based node ops for replace by deault. With/without this patch, if we enable repair based node ops, the results are the same.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+                ss.node_ops_update_heartbeat(ops_uuid);
+            } else if (req.cmd == node_ops_cmd::replace_done) {
+                slogger.info("replace[{}]: Marked ops done from coordinator={}", req.ops_uuid, coordinator);
+                ss.node_ops_done(ops_uuid);
```
I've noticed that `node_ops_done` doesn't fail if the op is not found. That may lead to consistency problems where the coordinator thinks the operation succeeded while it failed on some nodes due to watchdog abort.

Consider this scenario:
1. node A is the coordinator for the node op
2. node B is just a regular node 
3. streaming completes
4. node B missed the heartbeat so it rolled back to the previous topology
5. writes were made through B, and they went to the old topology
6. node A sends `replace_done`, which succeeds everywhere, even on node B
7. node A switches the cluster to the new topology via gossip, but writes made in step 5 will not be there (or with smaller CL)

asias replied:
In v2, I have already added  extra check about the uuids.

tgrabiec replied:
What if node B restarts after `replace_done` in step 6? It will revert to the old topology temporarily and run into the same issue, I think

asias replied:
If your example, if B missed the heartbeat and rolled back, B will abort the ops with the uuid, delete the host id, the replace_done cmd will FAIL. Node A will faill the ops because it fails at the sening replace_done process. 


If a node restarts after the finish step (node_ops_cmd::replace_done), it is fine because after that, the replacing node will set gossip status to NROMAL. When the node restarts, it will learn the replacing node is a NORMAL node in the cluster.

tgrabiec replied:
The replacing node sets the gossip status after sending replace_done requests. Node B can restart before it sees the replacing node jumping to NORMAL, because it happens asynchronously.

asias replied:
What's wrong with this case? Yes, Node B restarts before it knows the replacing node jumps to NORMAL, after it restarts  will  it learn the NORMAL status from the rest of the cluster. This is exactly what Node B is suppose to see.

tgrabiec replied:
The problem is that node B, after it restarts, may see and use the old topology for some time window, which can lead to lost writes or inconsistent reads.

asias replied:
Please note the old and new topology are the same after replace is done. I do not see how lost of writes or inconsistent reads would happen with v3. If you disagree, please give concrete examples.

asias commented:
Changes in v2:
- Added log before abort process  and change the abort msg to error level
- Added storage_service::node_ops_cmd_check to check if we know the ops uuid to make sure only one ops at a time and prevents restarted node to handle ops id it does not know. 
- Added gossiper::is_normal_or_shutdown helper
- Added gossiper::wait_alive helper
- Added log for ep in "Failed to send heartbeat" log

asias commented:
Changes in v3:
- Added gossiper::advertise_to_nodes to protect restarted node to mark replacing node as alive while not knowing the replacing node is replacing
- Avoid send abort cmd if it has done responded done cmd.

tgrabiec reviewed (changes requested):
v3 solves the restart problem for the duration of the replace operation.

I think it still doesn't solve the race which happens at the end when we commit the operation. After all nodes ACK `replace_done`, we no longer can detect restarts. The replacing will succeed. If a restart happens then, we will run into consistency problems, where the restarted node will use the old topology for a while until it sees the replaced node in NORMAL status via gossip.

I don't see how this can be solved other than having the relevant local state persistent and monotonic instead of in-memory only.  
Current system supports this partially. We persist normal node states in system.peers. There was a plan to extend this to all statuses. Refs https://github.com/scylladb/scylla/issues/6088. This would solve problems with restarts here, if we persist the hibernate status here.

I think that monotonic state is the right long term solution because it will enable the algorithm to proceed rather than abort. Node restart is like a temporary network partition.

asias commented:
> v3 solves the restart problem for the duration of the replace operation.
> 
> I think it still doesn't solve the race which happens at the end when we commit the operation. After all nodes ACK `replace_done`, we no longer can detect restarts. The replacing will succeed. If a restart happens then, we will run into consistency problems, where the restarted node will use the old topology for a while until it sees the replaced node in NORMAL status via gossip.

After the replacing is done, the replacing node is the ready to serve either read or write, it is technical the same as the node to be replaced. 

For example, n1,n2,n3 (127.0.0.3). n3 is the replacing node.

n1, n2 acks replace_done verb.

Then n1 restarts, before n1 knows n3 is in NORMAL status, it thinks 127.0.0.3(the old node) is DOWN and NORMAL. 

What is the problem?

Note, with this patch, during replacing, HIBERNATE state is not injected into the cluster.

asias commented on service/storage_service.cc:
```diff
@@ -1911,6 +1914,149 @@ future<> storage_service::decommission() {
+
+
+        // Step 4: Allow nodes in sync_nodes list to mark the replacing node as alive
+        _gossiper.advertise_to_nodes(sync_nodes).get();
```
Yes, this is better, we need two steps to capture and set though. Will integrate in v4.

asias commented on gms/gossiper.hh:
```diff
@@ -572,6 +572,10 @@ public:
+    // Check if a node is in NORMAL or SHUTDOWN status which means the node is
+    // part of the token ring from the gossip point of view and operates in
+    // normal status or was in normal status but is shutdown.
+    bool is_normal_or_shutdown(const inet_address& endpoint) const;
```
ok

asias commented:
Changes in v4:
- Rename is_normal_ring_member
- Sample gossip generation earlier

tgrabiec commented:
On Thu, Apr 1, 2021 at 2:42 AM Asias He ***@***.***> wrote:

> v3 solves the restart problem for the duration of the replace operation.
>
> I think it still doesn't solve the race which happens at the end when we
> commit the operation. After all nodes ACK replace_done, we no longer can
> detect restarts. The replacing will succeed. If a restart happens then, we
> will run into consistency problems, where the restarted node will use the
> old topology for a while until it sees the replaced node in NORMAL status
> via gossip.
>
> After the replacing is done, the replacing node is the ready to serve
> either read or write, it is technical the same as the node to be replaced.
>
> For example, n1,n2,n3 (127.0.0.3). n3 is the replacing node.
>
> n1, n2 acks replace_done verb.
>
> Then n1 restarts, before n1 knows n3 is in NORMAL status, it thinks
> 127.0.0.3(the old node) is DOWN and NORMAL.
>
> What is the problem?
>

I don't think there is a problem with consistency for the replace operation
anymore.

avikivity commented:
> > Looks good. But every time we enable repair based operations we revert the, later. What testing was done to make sure this doesn't happen this time?
> 
> This patch is more about the control plane. The data plane that uses repair to sync data does not change. I have tested with dtest. This patch does not solve any latency issues during replace.

Does it mean that the decision whether to use repair or not does not change (and it will still not be used by default?)

avikivity commented:
I'll merge, but really the credit for reviewing this should go to @tgrabiec.
